### ISSUE TITLE
Add support for annotations in the fluent-bit service

### ISSUE THREAD

**Issue by Kristian-ZH** [fluent/fluent-operator#654]
### Is your feature request related to a problem? Please describe.

Currently, the fluent-bit service supports only labels but not annotations. I think it will be good if we can add annotations also in it.

### Describe the solution you'd like

As the config options of the service are increasing, my idea is if we add a `service` section in the FluentBit resource and there to add all of the service configurations?

Also currently `MakeDaemonSet` and `MakeService` return Golang objects which will create a copy of the objects each time when the Operator reconciles. In the improvement mentioned above, this can be also changed and make both functions to return pointers.

**Comment by benjaminhuo** [comment]
@Kristian-ZH, @HarleyB123 already added a `ServiceAnnotation` field in https://github.com/fluent/fluent-operator/pull/647
What other service fields would you like to add? A separate Service section might be better than a single`ServiceAnnotation`

**Comment by benjaminhuo** [comment]
> Also currently `MakeDaemonSet` and `MakeService` return Golang objects which will create a copy of the objects each time when the Operator reconciles. In the improvement mentioned above, this can be also changed and make both functions to return pointers.

Agree

**Comment by Kristian-ZH** [comment]
> @Kristian-ZH, @HarleyB123 already added a ServiceAnnotation field in https://github.com/fluent/fluent-operator/pull/647

https://github.com/fluent/fluent-operator/blob/master/pkg/operator/fluent-bit-service.go#L29

The PR that you mentioned enables the Annotation for the SA, but I would like to have a support for Annotation on the Service.

Also, currently the Service uses the same Labels, name and so on as the FB Daemonset. What about if we do them separately.

For example to have a `service` section in the FluentBit Resource and there to add everything you need for the service as `name`, `labels`, `annotations` and so on...

**Comment by benjaminhuo** [comment]
> > @Kristian-ZH, @HarleyB123 already added a ServiceAnnotation field in #647
> 
> https://github.com/fluent/fluent-operator/blob/master/pkg/operator/fluent-bit-service.go#L29
> 
> The PR that you mentioned enables the Annotation for the SA, but I would like to have a support for Annotation on the Service.
> 
> Also, currently the Service uses the same Labels, name and so on as the FB Daemonset. What about if we do them separately.
> 
> For example to have a `service` section in the FluentBit Resource and there to add everything you need for the service as `name`, `labels`, `annotations` and so on...

Sorry, I mixed up Service and ServiceAccount.
You're right, a separate service section in FluentBit CR looks good.

**Comment by benjaminhuo** [comment]
> For example to have a `service` section in the FluentBit Resource and there to add everything you need for the service as `name`, `labels`, `annotations` and so on...

There should be default values that are the same as before if a service section is not set for backward compatibility.

**Comment by Kristian-ZH** [comment]
Sure, I can do this

**Comment by benjaminhuo** [comment]
Currently, the service name and labels are the same as the FluentBit CR, and there're no annotations.
You need to set annotations for service and specify service names and labels that differ from the FluentBit CR, right?

**Comment by Kristian-ZH** [comment]
One option will be if I modify the FluentBit CR like this:
```
  kind: FluentBit
  metadata:
    name: fluent-bit
    namespace: test
  spec:
    fluentBitConfigName: fluent-bit-config
    image: kubesphere/fluent-bit:v2.0.6
    ....
    service:
      name: ...
      labels...
      annotations....
```

or smth. like this. Another idea is to create a dedicated CR only for the Service.

What do you think?

**Comment by benjaminhuo** [comment]
> or smth. like this. Another idea is to create a dedicated CR only for the Service.

I think no need to create a dedicated CR, the method you mentioned is ok

**Comment by Kristian-ZH** [comment]
Okay then will follow this approach

**Comment by ialidzhikov** [comment]
@Kristian-ZH , @benjaminhuo can this issue be closed? I see that https://github.com/fluent/fluent-operator/pull/657 was merged.

**Comment by benjaminhuo** [comment]
> @Kristian-ZH , @benjaminhuo can this issue be closed? I see that #657 was merged.

Yes, sure
